### ISSUE TITLE
开发者您好，关于Wavemlp的疑问

### ISSUE THREAD

**Issue by jczzp** [huawei-noah/CV-Backbones#90]
尊敬的开发者您好：
       最近学习您的Wavemlp代码，结合论文发现在token融合时代码与论文中的描述似乎有出入，具体涉及wavemlp.py中87,88,111，112行，这部分代码摘抄如下：
        self.tfc_h = nn.Conv2d(2*dim, dim, (1,7), stride=1, padding=(0,7//2), groups=dim, bias=False) 
        self.tfc_w = nn.Conv2d(2*dim, dim, (7,1), stride=1, padding=(7//2,0), groups=dim, bias=False)
        x_h=torch.cat([x_h*torch.cos(theta_h),x_h*torch.sin(theta_h)],dim=1)
        x_w=torch.cat([x_w*torch.cos(theta_w),x_w*torch.sin(theta_w)],dim=1)
        据我所知，torch.cat拼接操作沿着C方向拼接时，以32通道x_h拼接为例，第一项乘cos为实部，后面为虚部，经过拼接后通道数将翻倍，但其顺序是32通道的实部+32通道虚部，之后进行分组卷积时似乎并不是如论文中所述“每一个Token的实部与虚部融合”，更像是所有Token的实部两两融合后，所有Token的虚部再两两融合。
        问题的关键是torch.cat操作中似乎并不是交替拼接，导致最终拼接后的数据呈现“所有实部与所有虚部”，而不是32组“实部与虚部”的组合，这导致后续分组卷积时似乎并未正确融合每一个Token的实部与其对应的虚部。
        由于自身对Pytorch理解尚浅，上述描述如有错误请您不吝赐教，希望您能够解答关于论文中公式9在代码中的实现，感谢您的关注，期待您的回复！

**Comment by jczzp** [comment]
@iamhankai @ggjy @yitongh @huawei-noah-admin

**Comment by yehuitang** [comment]
Hi! It is a good question. Considering that there are linear layers along the channel dimension after the token aggregation, the different implements you mentioned are almost equivalent in practice.

**Comment by jczzp** [comment]
感谢您的回复，基于论文中的思想，我认为可以将代码如下修改，希望您指导一下我的想法是否正确：
             主要思想：取消torch.cat拼接操作，先通过欧拉公式展开式分别计算实部虚部，将其通过两个独立的group conv做token融合，由于输入输出通道数与group均相同，此时group conv等价于depthwise conv，只在特征图维度做空间融合而没有通道融合，符合token-FC思想）。随后将token融合后的结果相加，这样应该满足论文中实部虚部对应融合的公式，能够确保融合对象是每一个token的实部与虚部，代码如下：
             self.tfc_h = nn.Conv2d(dim, dim, (1,7), stride=1, padding=(0,7//2), groups=dim, bias=False) 
             self.tfc_w = nn.Conv2d(dim, dim, (7,1), stride=1, padding=(7//2,0), groups=dim, bias=False)
             x_hr=x_h*torch.cos(theta_h)
             x_hi=x_h*torch.sin(theta_h)
             x_wr=x_w*torch.cos(theta_w)
             x_wi=x_w*torch.sin(theta_w)
             h = self.tfc_h(x_hr)+ self.tfc_h(x_hi)
             w = self.tfc_w(x_wr)+ self.tfc_w(x_wi)
             请批评指正上述思路是否正确，期待与您进一步沟通，再次感谢！

**Comment by jczzp** [comment]
@yehuitang
